src/main/java/io/cryostat/configuration/CredentialsManager.java

```diff
@@ -67,137 +60,61 @@
-                                        return 0;
-                                    }
-                                })
-                        .reduce(Math::max)
-                        .orElse(0);
-    }
-
     public int addCredentials(String matchExpression, Credentials credentials)
-            throws IOException, MatchExpressionValidationException {
+            throws MatchExpressionValidationException {
         matchExpressionValidator.validate(matchExpression);
-        Path destination = getPersistedPath(nextId);
-        fs.writeString(
-                destination,
-                gson.toJson(new StoredCredentials(matchExpression, credentials)),
-                StandardOpenOption.WRITE,
-                StandardOpenOption.CREATE,
-                StandardOpenOption.TRUNCATE_EXISTING);
-        fs.setPosixFilePermissions(
-                destination,
-                Set.of(PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_WRITE));
+        StoredCredentials saved = dao.save(new StoredCredentials(matchExpression, credentials));
         emit(CredentialsEvent.ADDED, matchExpression);
-        return nextId++;
+        return saved.getId();
     }
 
-    public int removeCredentials(String matchExpression)
-            throws IOException, MatchExpressionValidationException {
+    public int removeCredentials(String matchExpression) throws MatchExpressionValidationException {
+        // TODO refactor this to do a proper query on the DAO
         matchExpressionValidator.validate(matchExpression);
-        for (String pathString : this.fs.listDirectoryChildren(credentialsDir)) {
-            Path path = credentialsDir.resolve(pathString);
-            try (BufferedReader br = fs.readFile(path)) {
-                StoredCredentials sc = gson.fromJson(br, StoredCredentials.class);
-                if (Objects.equals(matchExpression, sc.getMatchExpression())) {
-                    Path filenamePath = path.getFileName();
-                    if (filenamePath == null) {
-                        throw new IllegalStateException(path.toString());
-                    }
-                    String filename = filenamePath.toString();
-                    fs.deleteIfExists(path);
-                    return Integer.parseInt(filename);
-                }
+        for (StoredCredentials sc : dao.getAll()) {
+            if (Objects.equals(matchExpression, sc.getMatchExpression())) {
+                int id = sc.getId();
+                dao.delete(id);
+                emit(CredentialsEvent.REMOVED, matchExpression);
+                return id;
             }
         }
-        throw new FileNotFoundException();
+        throw new IllegalArgumentException();
     }
 
-    public Credentials getCredentialsByTargetId(String targetId)
-            throws JsonSyntaxException, JsonIOException, IOException, ScriptException {
+    public Credentials getCredentialsByTargetId(String targetId) throws ScriptException {
```

Comment:
One problem that I was sometimes running into was that during my work on the Archives and JvmMapping PRs, on the `MetadataManager` startup, the code tests connections to targets to see if active recordings are still valid and uses `Credentials` to access them through JMX. But sometimes since the task was async, the `platformClient` is not started fast enough to `listDiscoverableServices` and that's why sometimes there are `Credential` errors. So it seems the `DiscoveryStorage` should be loaded first but the reason I changed the verticles to start in order was because the `TargetJvmDiscovery` ws notifications should happen after the `MetadataManager` startup. Maybe to solve this, if possible add a different `getCredentialsByTargetId` that only uses the Dao to query without `listDiscoverableServices` but not having to pass an entire `serviceRef`?

Reply:
I'm not sure if I follow. I think this would mean adding a method that basically checks for credentials defined with a match expression specifically like `target.connectUrl == some-specific-targetId`? That's conceptually different from what the current implementation does, which is looking for a target that matches the targetId given and then checking if there are any matchExpression-based credentials that fit the corresponding ServiceRef - and they can be matching based on any property of the ServiceRef, not only the connectUrl. Without using a ServiceRef object or listDiscoverableServices or any other information from the DiscoveryStorage I'm not sure how we can query for any stored credentials that match a target by connectUrl alone. It would only work for certain credentials that are only defined for one specific target.

If I'm understanding the problem correctly it seems that the root cause is that the MetadataManager wants to do its pruning immediately at startup based on what it can see on disk, but it needs to cross-reference those against discovered targets from the DiscoveryStorage. Depending on what discovery plugins are active, the DiscoveryStorage may end up reloading an empty state and waiting for async network processes to occur before repopulating the database - this is always what will happen with the DefaultPlatformClient/JDP for example. If you're running in OpenShift with the KubernetesApiPlatformClient we can mitigate this by ensuring the storage is repopulated with a startup Kubernetes API query so that the storage isn't considered as "started" until after we have loaded target data into it. This isn't currently done but it should be easy enough to add.

Another angle on a fix would be for the MetadataManager to check what data is has on disk and queue up jobs to validate that data. When a matching target does appear then the corresponding job is executed and we connect to the target to check if the recordings we expect are still there. If the target never appears within some time interval then we assume it's never going to and we clear the data instead. This way the metadata prune is a little more lazily evaluated instead of eagerly and it is more tolerant of delayed target reappearances, which are never entirely avoidable.

Reply:
Anyway, that was very long-winded. We can have a meeting to talk through those details and I can try to draw up some illustrations to depict the data flow and dependencies as I'm picturing it.

Reply:
> I'm not sure if I follow. I think this would mean adding a method that basically checks for credentials defined with a match expression specifically like `target.connectUrl == some-specific-targetId`? That's conceptually different from what the current implementation does, which is looking for a target that matches the targetId given and then checking if there are any matchExpression-based credentials that fit the corresponding ServiceRef - and they can be matching based on any property of the ServiceRef, not only the connectUrl. Without using a ServiceRef object or listDiscoverableServices or any other information from the DiscoveryStorage I'm not sure how we can query for any stored credentials that match a target by connectUrl alone. It would only work for certain credentials that are only defined for one specific target.
> 
Ah, I see then I was thinking about this differently than how it actually is.

> If I'm understanding the problem correctly it seems that the root cause is that the MetadataManager wants to do its pruning immediately at startup based on what it can see on disk, but it needs to cross-reference those against discovered targets from the DiscoveryStorage. Depending on what discovery plugins are active, the DiscoveryStorage may end up reloading an empty state and waiting for async network processes to occur before repopulating the database - this is always what will happen with the DefaultPlatformClient/JDP for example. Other mechanisms like Custom Targets and KubernetesApiPlatformClient are able to load their plugin information to the discovery database on startup (Custom Targets by just reloading what was already in the DB, Kubernetes by performing a k8s API query for Endpoints and building the subtree) so I think they're already less susceptible to this class of problem.
> 
> Another angle on a fix would be for the MetadataManager to check what data is has on disk and queue up jobs to validate that data. When a matching target does appear then the corresponding job is executed and we connect to the target to check if the recordings we expect are still there. If the target never appears within some time interval then we assume it's never going to and we clear the data instead. This way the metadata prune is a little more lazily evaluated instead of eagerly and it is more tolerant of delayed target reappearances, which are never entirely avoidable. This way the MetadataManager and CredentialsManager can both start up early, the MetadataManager does not have an immediate direct coupling/dependency on credentials being available (which can sometimes require the DiscoveryStorage to be ready too), and then we can start the DiscoveryStorage after. Once the DiscoveryStorage starts emitting notifications that's what triggers the MetadataManager to start validating its findings, but at this point the DiscoveryStorage is obviously already ready, which means it's also ready for the CredentialsManager to reference it. I think any kind of discovery plugin, even ones that "trickle" new target updates like JDP, would work in this scheme since the metadata validation is done on-demand as each target appears.
> 

That makes sense and solves the problem, sure that can be changed in a new issue I will make.

> The last hole in this scheme is that for platform mechanisms that don't "trickle" and simply repopulate fully on startup (like custom targets or kubernetes), you don't get any notifications for those targets. If the MetadataManager does not start until after the DiscoveryStorage has completed startup though, then this case should be handled by just having the MetadataManager call `listDiscoverableServices` on its own startup and hook that into the same validation job queue processing as the target discovery events go into as described above.

The MetadataManager completes before the DiscoveryStorage starts however, so this would not work unfortunately. Seems like a way to do this would just to send a callback to the MetadataManager when the DiscoveryStorage is ready to do the queue processing job again, but I'm wondering about a way to do this without, like you said, a direct dependency on credentials and the DiscoveryStorage to be ready.

Reply:
> Anyway, that was very long-winded. We can have a meeting to talk through those details and I can try to draw up some illustrations to depict the data flow and dependencies as I'm picturing it.

I think it makes sense to me, if something gets too confusing I will let you know, thanks!

Reply:
> The MetadataManager completes before the DiscoveryStorage starts however, so this would not work unfortunately. 

Does it need to? If the metadata validation was done lazily then the MetadataManager could be started after the DiscoveryStorage. Both instances could be constructed, with the MetadataManager registering a listener for target discovery events and add them to its queue of observed targets. Then the DiscoveryStorage can be started and allowed to complete, then the MetadataManager can be started. The MetadataManager scans its storage and creates a map of `connectUrl` to metadata-to-be-validated, then calls `listDiscoverableServices` once and adds those results to the queue of observed targets. Then it starts a task that takes elements from the observed targets queue and checks if it matches an entry in the validation map. If it does, it connects to the target and checks if the stored metadata is still accurate, performing the migration as needed and removing the validation map entry. If the target from the queue doesn't match anything in the validation map just skip it. Then, separately, drop entries from the validation map if they become too old and stale.

I think key data structures to use here are probably a Caffeine cache for the validation map and something like a BlockingQueue for the observed targets queue. The target discovery callback would place elements into queue with `add()`, and the metadata validation tasks would remove elements from the queue using a loop and `take()`. Once the validation map is empty the validation task can terminate itself.

Reply:
Right, I was thinking about on startup, at least on the JDP side of things, the BuiltInDiscovery which is part of the DiscoveryStorage is a verticle which sends notifications that the MetadataManager wont be able to see if it starts afterwards, but I see that if we just have the MetadataManager start after the DiscoveryStorage, then a call to `listDiscoverabeServices` should cover those notifications we missed anyways. 

> I think key data structures to use here are probably a Caffeine cache for the validation map and something like a BlockingQueue for the observed targets queue. The target discovery callback would place elements into queue with add(), and the metadata validation tasks would remove elements from the queue using a loop and take(). Once the validation map is empty the validation task can terminate itself.

Sure I'll look into it, looks like its back in the saddle with these types of issues. :-)

Reply:
Makes sense. Yea, the MetadataManager can register a target discovery listener and begin receiving updates before it itself actually goes through Verticle startup. That plus the `listDiscoverableServices` upon startup should ensure that the MetadataManager doesn't miss any targets.

Reply:
Oh, one more thought along that line. There might be a good way to use the Vertx eventbus here to do the target discovery communication between storage and metadata manager. Rather than forking off a separate worker thread or something, the metadata manager could just listen for discovery events over the event bus, and the storage can publish those events on the bus. This way the verticles' worker threads handle that async job nicely already. This might change the scheme of registering an early listener hook before starting the verticle however - it might mean the metadata manager actually has to start first so it can register its event bus listener before the storage starts and potentially begins to publish events. The storage could also publish a message indicating that its startup has completed, and the metadata manager can listen for that message and use that to trigger calling `listDiscoverableServices`.

Reply:
Yeah makes sense, I'll try stuff out and see what works and what doesn't, thanks for the ideas!